internal/unsafecast/unsafecast.go

```diff
@@ -81,7 +81,11 @@ func Slice[To, From any](data []From) []To {
 
 // Bytes constructs a byte slice. The pointer to the first element of the slice
 // is set to data, the length and capacity are set to size.
+// if the data passed points to null, and empty byte array of the given size and cap is returned.
 func Bytes(data *byte, size int) []byte {
+	if data == nil {
+		return make([]byte, size, size)
+	}
```

Comment:
This doesn't seem like the right place to handle this condition; the purpose of this package is to provide casting between Go types, creating a new value modifies the semantics of the function under some conditions.

Were you able to determine the code paths that led to this condition? I would recommend we handle this condition at the call side where a `nil` data was being passed.

Reply:
Given the abstractions I found it a bit difficult to get to the root of the problem, I have some guesses though. Unfortunately the data I'm migrating is considered sensitive and classified by the company, so I can not share the file with you, but here I got a stack trace of the nil pointer which I hope will be of help to find the root cause.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x6b302a]

goroutine 1 [running]:
encoding/binary.littleEndian.Uint32(...)
        /snap/go/10389/src/encoding/binary/binary.go:81
github.com/parquet-go/parquet-go.makeInt96(...)
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/value.go:800
github.com/parquet-go/parquet-go.(*Value).int96(...)
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/value.go:461
github.com/parquet-go/parquet-go.convertInt96ToInt32({0x0, 0x4a5d, 0xfe, 0x1, 0x0, 0xfffe})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/convert.go:705 +0x12a
github.com/parquet-go/parquet-go.int32Type.ConvertValue({}, {0x0, 0x4a5d, 0xfe, 0x1, 0x0, 0xfffe}, {0x8559f8?, 0xb76e00?})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/type.go:421 +0x351
github.com/parquet-go/parquet-go.(*dateType).ConvertValue(0x0?, {0x0, 0x4a5d, 0xfe, 0x1, 0x0, 0xfffe}, {0x8559f8?, 0xb76e00?})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/type.go:1763 +0x23b
github.com/parquet-go/parquet-go.convertToType.func1({0xc04e337260, 0x1, 0xffff000001fe?})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/convert.go:86 +0xf6
github.com/parquet-go/parquet-go.multiConversionFunc.func1({0xc04e337260, 0x1, 0x3c})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/convert.go:143 +0x6b
github.com/parquet-go/parquet-go.(*conversion).Convert(0xc00007fe00, {0xc051cd8000, 0x20002, 0x21bdfa?})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/convert.go:202 +0x54b
github.com/parquet-go/parquet-go.(*convertedRows).ReadRows(0xc04e360690, {0xc051cd8000?, 0x329e000?, 0x21bdfa})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/convert.go:561 +0x72
github.com/parquet-go/parquet-go.(*reader).ReadRows(0xc045118f10, {0xc051cd8000?, 0x21bdfa?, 0x21bdfa?})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/reader.go:526 +0xc2
github.com/parquet-go/parquet-go.(*Reader).ReadRows(0xc045118f00, {0xc051cd8000, 0x21bdfa, 0x21bdfa})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/reader.go:443 +0x66
github.com/parquet-go/parquet-go.(*GenericReader[...]).readRows(0x852000, {0xc04f9fe000, 0x21bdfa, 0x0})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/reader.go:160 +0xfd
github.com/parquet-go/parquet-go.(*GenericReader[...]).Read(...)
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/reader.go:114
github.com/parquet-go/parquet-go.Read[...]({0x84d060, 0xc007c81cc8}, 0x4dbc8c, {0x0?, 0x7b5880?, 0xc025d6a1a0?})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/parquet.go:36 +0x126
github.com/parquet-go/parquet-go.ReadFile[...]({0xc00053350c, 0xc02094ded8?}, {0x2?, 0x2?, 0x749260?})
        /root/go/pkg/mod/github.com/!mahdi-zarei/parquet-go@v0.18.1-0.20231017111659-9fd16ddf62d0/parquet.go:63 +0x12c
```

the trace is run on a version which has its int96 function modified, the modified function is

```
func makeInt96(bits []byte) (i96 deprecated.Int96) {
	if len(bits) < 12 {
		bits = append(bits, make([]byte, 12-len(bits))...)
	}
	return deprecated.Int96{
		2: binary.LittleEndian.Uint32(bits[8:12]),
		1: binary.LittleEndian.Uint32(bits[4:8]),
		0: binary.LittleEndian.Uint32(bits[0:4]),
	}
}
```

as you can see the nil pointer can not occur in a normal situation, so the only conclusion I could draw is that we had misused the Unsafe package.
I will keep looking for the root cause, any hints would be greatly appreciated.

Reply:
I guess I found the problem, the ConvertValue function called like t1.ConvertValue(v, t2) tries to convert v of kind t2 to t1, and it was being called in the inverse manner, causing really odd behaviour. After [fix wrong conversion](https://github.com/parquet-go/parquet-go/pull/78/commits/9247860cd6e79fb97561535eedbd18c21fee01f9) it correctly tries to convert from srcType to targetType. If you agree with this, then I can revert my changes on unsafe.ByteArray and get the PR ready.

Reply:
I removed the change to unsafe.Bytes.